src/core/ComboBox/ComboBox.tsx

```diff
@@ -0,0 +1,337 @@
+import { Popover, PopoverProps } from '../utils/Popover';
+import { CommonProps } from '../utils/props';
+import SvgCaretDownSmall from '@itwin/itwinui-icons-react/cjs/icons/CaretDownSmall';
+import SvgCaretUpSmall from '@itwin/itwinui-icons-react/cjs/icons/CaretUpSmall';
+
+export type ComboBoxProps<T> = {
+  /**
+   * Array of options that populate the dropdown list.
+   */
+  options: SelectOption<T>[];
+  /**
+   * Controlled value of ComboBox.
+   */
+  value?: T;
+  /**
+   * Callback fired when selected value changes.
+   */
+  onChange?: (value: T | null) => void;
+  /**
+   * Function to customize the default filtering logic.
+   */
+  filterFunction?: (
+    options: SelectOption<T>[],
+    inputValue: string,
+  ) => SelectOption<T>[];
+  /**
+   * Native input element props.
+   */
+  inputProps?: Omit<React.ComponentPropsWithoutRef<'input'>, 'size'>;
+  /**
+   * Props to customize dropdown menu behavior.
+   */
+  dropdownMenuProps?: PopoverProps;
+  /**
+   * Message shown when no options are available (after filtering).
+   * @default 'No options found'
+   */
+  emptyStateMessage?: string;
+} & Omit<CommonProps, 'title'>;
+
+/**
+ * ComboBox component that allows typing a value to filter the options in dropdown list.
+ * Values can be selected either using mouse clicks or using the Enter key.
+ * @example
+ * <ComboBox
+ *   options={[
+ *     { label: 'Item 1', value: 1 },
+ *     { label: 'Item 2', value: 2 },
+ *     { label: 'Item 3', value: 3 },
+ *   ]}
+ *   onChange={() => {}}
+ * />
+ */
+export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+  const {
+    options,
+    value,
+    onChange,
+    filterFunction,
+    className,
+    inputProps,
+    dropdownMenuProps,
+    emptyStateMessage = 'No options found',
+    ...rest
+  } = props;
+
+  // Generate a stateful random id if not specified
+  const [id] = React.useState(() => props.id ?? `iui-${nanoid(10)}`);
+
+  useTheme();
+
+  /** Generates an memoized id for an option, given the index from original list */
+  const getOptionId = React.useCallback(
+    (index: number) =>
+      options[index].id ??
+      `${id}-option${options.findIndex(
+        ({ value }) => value === options[index].value,
+      )}`,
+    [options, id],
+  );
+
+  const inputRef = React.useRef<HTMLInputElement>(null);
+  const menuRef = React.useRef<HTMLUListElement>(null);
+  const [isOpen, setIsOpen] = React.useState(false);
+
+  // Set min-width of menu to be same as input
+  const [minWidth, setMinWidth] = React.useState(0);
+  React.useEffect(() => {
+    if (inputRef.current) {
+      setMinWidth(inputRef.current.offsetWidth);
+    }
+  }, [isOpen]);
+
+  const [filteredOptions, setFilteredOptions] = React.useState(options);
+  const [focusedIndex, setFocusedIndex] = React.useState<number>(-1);
+
+  // Maintain internal selected value state synced with `value` prop
+  const [selectedValue, setSelectedValue] = React.useState<T | undefined>();
```

Comment:
Just started to think, do we really want to save the selected value internally? Shouldn't we leave this to the user?

Reply:
How would it help? It's much nicer to support both